### ISSUE TITLE
Security: Support SSL from API configuration 

### ISSUE THREAD

**Issue by zeeiyerWork** [precog/quasar#2390]
(TBD: Also requires SD ticket). 
@jdegoes: @sellout needed more clarification on what this line item in the epic is about. 
Is it about allowing quasar to query Mongodb over SSL -> as stated in related Mongodb [Documentation for the scala driver.](http://mongodb.github.io/mongo-scala-driver/1.1/reference/connecting/ssl/?_ga=2.192403050.1390275891.1498176459-1217746313.1493763568)
And the Slamdata ticket would be to specify this when mounting a specific Mongodb?

**Comment by jdegoes** [comment]
The mounting API for mongo DB needs to accept SSL configuration information to connect to the mongo database. Presently, this has to be done by specifying An esoteric combination of properties for the Java/Scala driver used by Quasar.  The feature needs to be entirely visual and driven by the mount dialogue, not something users have to muck about with property files in order to configure.

**Comment by jdegoes** [comment]
> And the Slamdata ticket would be to specify this when mounting a specific Mongodb?

Yes, these details should be specified on the mount dialog for MongoDB.

**Comment by zeeiyerWork** [comment]
[Slamdata documentation](http://docs.slamdata.com/en/latest/administration-guide.html?highlight=SSL#mongodb) claims that this is already supported (i.e. the ability to request for Slamdata to I/o against Mongodb via SSL) - presume that Quasar (and old Mongo) consequently honor it -> 
![image](https://user-images.githubusercontent.com/24783629/27496357-77f99152-580a-11e7-85e5-d165cc577d9d.png)
Hence the confusion... @sellout, please comment on whether this does work in old-mongo - have not tried it out personally.

**Comment by jdegoes** [comment]
As far as I know, SSL connections to MongoDB have no API and no visual configuration. /cc @damonLL

**Comment by damonLL** [comment]
@jdegoes: correct. No real API for configuring connections. You supply the connection URI to initially connect, which is where you specify all of the options (timeout, SSL, read preference, etc)

To achieve SSL connections with MongoDB, we currently must do two specific things:

* Provide the ‘ssl‘ configuration parameter to mongo connector options (thus to the connection URI) as it shows in the docs.
* Properly setup SSL certificates for the JVM with the Java key store since we are using the Mongo driver.

This doesn't lend itself well to UI configuration since it needs to be setup before JVM launch (from everything I've read and experimented with)

The second step is usually troublesome for most people. MongoDB does not provide a plethora of examples and unfortunately the process always varies (at least slightly) depending on the cert provider as well as the MongoDB hosting provider.

Let me know if this needs additional clarification, happy to hop on a call.

**Comment by jdegoes** [comment]
Part (1) we should put into the UI and make it simple.

Part (2) we should bake into Quasar and drive it by API so it can be visual.

I doubt doing before JVM launch is the only way to do it — likely there's another way. If not, the Quasar could always restart itself (worst case).

**Comment by zeeiyerWork** [comment]
Given:
>>> Provide the ‘ssl‘ configuration parameter to mongo connector options (thus to the connection URI) as **it shows in the docs.**

I take it, this ticket is 
- About verifying that Quasar respects the SSL configuration.
- Trying to see if something could be done to simplify setup for SSL certificates for the Slamdata JVM? (i.e. some UI on the frontend to allow for upload of a SSL certificate to the Slamdata backend). For Quasar (this ticket) it is about APIs to accept a SSL certificate and register it? And validate requests against it? 

Is this high priority (i.e. a customer waiting on it) or a nice to have?

**Comment by jdegoes** [comment]
A number of customers would use this feature immediately and numerous other users have encountered problems with cloud MongoDB providers that require SSL to be configured properly (most who try and encounter difficulties probably don't persevere long enough to figure a way around them).

(1) Quasar needs to support SSL connections to MongoDB via API,  without requiring a user to set system properties of the JVM; (2) SlamData needs to allow visual specification of SSL connection parameters that would otherwise be specified as Java system properties as described in the MongoDB documentation for the Java client library.
